### ISSUE TITLE
Ambiguous set_header behavior depends on http/2

### ISSUE THREAD

**Issue by tekuchev** [haproxy/haproxy#1388]
### Detailed Description of the Problem

The header with URI (`%HP`), defined via `http-request set-header` which send to backend, changes depending the protocol on frontend:

If request via http1.1, var `%HP` has only path.
If request  via http/2, var `%HP` has full URL: schema, host and path. 

The problem produced on versions 2.4.3 and 2.2.15. Variable %HU also affected with this problem.
This problem doesn't exists on haproxy-2.0.24.

### Expected Behavior

1. Variable's schema shouldn't depend on frontend protocol.
2. Content of variables should be described in doc.

### Steps to Reproduce the Behavior

php-script to echo header `x-fullurl` on backend:
```
<?php 
echo $headers['x-fullurl'];
echo "\n";
?>
```

Requests:
```
###  Normal behavior: %HP has only path:
$ curl --http1.1 -k https://example.com/print_header.php
/print_header.php

### Strange: %HP has protocol, host and path:
$ curl --http2 -k https://example.com/print_header.php
https://example.comf/print_header.php  

```

### What is your configuration?

```haproxy
global
        log log.local local6
        tune.ssl.default-dh-param 2048
        tune.bufsize 65536
        ssl-default-bind-options ssl-min-ver TLSv1.2
        daemon
        pidfile /var/run/haproxy.pid

defaults
        log global
        timeout connect 4000
        timeout client-fin 9000
        timeout http-keep-alive 200
        timeout client 120000
        timeout server 240000
        errorfile 503 /etc/haproxy/503.html

frontend f_name
        mode http
        bind 10.0.10.10:80 transparent
        bind 10.0.10.10:443 transparent alpn h2,http/1.1 ssl crt-list lists/fe_10.0.10.10_certs.list
        http-request replace-value host (.*):.* \1
        http-request set-var(req.backend) hdr(host),lower,map(lists/fe_10.0.10.10_domains.map,error_503)
        http-request add-header X-BACKEND %[var(req.backend)]
        http-request set-header x-fullurl %HP
        use_backend b_%[var(req.backend)]

backend b_name
        mode http
        server server1 10.0.10.10:80
        http-request set-header X-Forwarded-For %[src]
        http-request set-header X-SSL yes if { ssl_fc }
```

### Output of `haproxy -vv`

```plain
Linux hf1.local. 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux


HAProxy version 2.4.3-4dd5a5a 2021/08/17 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.3.html
Running on: Linux 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_OPENSSL=1 USE_SLZ=1 USE_TFO=1 USE_SYSTEMD=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 4.8.5 20150623 (Red Hat 4.8.5-44)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by wtarreau** [comment]
This is expected. It's one major difference between HTTP/1 and HTTP/2, HTTP/2 sends absolute URIs while HTTP/1 mostly sends origin URIs. Note that an HTTP/1 request sent to a proxy will be absolute and will have the same format as the HTTP/2 one. And absolute requests are valid in HTTP/1 when sent to servers, it's just that clients do not use them by default.

Could you please describe what you need ? (not what you'd expect from the existing tag). I suspect that you'd need some absolute URIs that are reconstructed from origin requests when needed, but I'd prefer to be sure. A long time ago it would have been a great pain to produce that but I think that nowadays it's something reasonable to add if that can be useful.

**Comment by wtarreau** [comment]
By the way I agree that more examples in the doc could help.

**Comment by tekuchev** [comment]
> Could you please describe what you need ?

In production we use this config to block unwanted URLs:

```
http-request set-header x-fullurl %[req.hdr(host),lower]%[url]
acl blocked_url req.fhdr(x-fullurl) -m reg -f lists/blocked_urls.list
use_backend b_blocked_url if blocked_url  # backend returns 503 error
```

The `%[url]` variable has same behavior and after upgrade from 2.0 to 2.2/2.4  blocked URLs became available via https. That was unexpected :-)
I already found  how do this in 2.4.3 using `%HPO` or  `%[path]` variables, so I made bug report because behavior of `%HP` variable  was changed, it is unclear and not described in docs.

**Comment by wtarreau** [comment]
I see. It's not really the variable itself which has changed but the compatibility matrix between the HTTP versions as spoken at the edge and the internal representation. We used not to keep the full URI in HTTP/2 before, but it was a real problem that made everything behave differently between HTTP/1 and HTTP/2, so we had to address it (and by the way HTTP is now heading into the direction of protocols that convey the full URI). In your case you noticed a difference in HTTPS because nowadays most HTTPS clients are browsers and they all support HTTP/2, thus they bring in the full URI while in HTTP/1 they do not.

I think that I should improve the HTTP introduction in the config manual, to explicitly mention differences between HTTP/1 and HTTP/2 that are visible to the user in the configuration. This way it would be easier in the various places where there is a possible difference to refer to this section to warn against possibly unexpected cases.
